### PR TITLE
Nuspec, packages.config, and package reference support

### PR DESCRIPTION
**Pull request by zlav**
This PR adds support for parsing `.nuspec`, `packages.config`, and package reference files for dependencies.

### CODE REVIEW

File: src/Strategy/NuGet/Nuspec.hs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+analyze :: Members '[ReadFS, Error ReadFSErr] r => BasicFileOpts -> Sem r (Graphing Dependency)
+analyze BasicFileOpts{..} = do
+      contents <- readContentsBS targetFile
+      let nuspecGroups = parseNuspec =<< XML.parseXMLDoc contents
+      
+      case nuspecGroups of
+        Just gs -> pure $ buildGraph gs
+        Nothing -> undefined
```
Comment (cnr) [inline src/Strategy/NuGet/Nuspec.hs]: `undefined` 😱

File: src/Strategy/NuGet/Nuspec.hs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  , strategyAnalyze = analyze
+  , strategyModule = parent . targetFile
+  , strategyOptimal = NotOptimal
+  , strategyComplete = NotComplete
+  }
+
+analyze :: Members '[ReadFS, Error ReadFSErr] r => BasicFileOpts -> Sem r (Graphing Dependency)
+analyze BasicFileOpts{..} = do
+      contents <- readContentsBS targetFile
+      let nuspecGroups = parseNuspec =<< XML.parseXMLDoc contents
```
Comment (cnr) [inline src/Strategy/NuGet/Nuspec.hs]: can we extract this pattern into helpers in `ReadFS`? -- similar to json/parser/yaml

File: src/Strategy/NuGet/Nuspec.hs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  childByName :: String -> XML.Element -> Maybe XML.Element
+  childByName name = XML.filterChildName (\elName -> XML.qName elName == name)
+
+  childrenByName :: String -> XML.Element -> [XML.Element]
+  childrenByName name = XML.filterChildrenName (\elName -> XML.qName elName == name)
+
+buildGraph :: [Group] -> Graphing Dependency
+buildGraph project = unfold direct (const []) toDependency
+    where
+    direct = foldr (\x y -> y ++ dependencies x) [] project
```
Comment (cnr) [inline src/Strategy/NuGet/Nuspec.hs]: `concatMap :: (a -> [b]) -> [a] -> [b]` codifies this pattern -- which might be a little easier to understand later. Let's use that instead

File: src/Strategy/NuGet/PackagesConfig.hs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  , strategyComplete = NotComplete
+  }
+
+analyze :: Members '[ReadFS, Error ReadFSErr] r => BasicFileOpts -> Sem r (Graphing Dependency)
+analyze BasicFileOpts{..} = do
+      contents <- readContentsBS targetFile
+      let packagesConfig = parsePackagesConfig =<< XML.parseXMLDoc contents
+      case packagesConfig of
+        Just gs -> pure $ buildGraph gs
+        Nothing -> undefined
```
Comment (cnr) [inline src/Strategy/NuGet/PackagesConfig.hs]: `undefined` again 😱

File: src/Strategy/NuGet/PackagesConfig.hs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  , strategyAnalyze = analyze
+  , strategyModule = parent . targetFile
+  , strategyOptimal = NotOptimal
+  , strategyComplete = NotComplete
+  }
+
+analyze :: Members '[ReadFS, Error ReadFSErr] r => BasicFileOpts -> Sem r (Graphing Dependency)
+analyze BasicFileOpts{..} = do
+      contents <- readContentsBS targetFile
+      let packagesConfig = parsePackagesConfig =<< XML.parseXMLDoc contents
```
Comment (cnr) [inline src/Strategy/NuGet/PackagesConfig.hs]: Can we extract this pattern to a helper in `ReadFS`? similar to json/parsec/yaml. We can use that helper in both of these strategies

File: hscli.cabal [hunk 6]
```diff
@@ -75,6 +75,7 @@ common deps
     , vector                       ^>=0.12.0.3
     , yaml                         ^>=0.11.1
     , yarn-lock                    ^>=0.6.2
+    , xml
```
Comment (cnr) [inline hscli.cabal]:
```suggestion
    , xml                          ^>=1.3.14
```
(adding the version bound)

File: src/Strategy/NuGet/Nuspec.hs (hunk 3 above)
Comment (zlav) [reply src/Strategy/NuGet/Nuspec.hs]: Didn't know about that, good suggestion

**Pull request merged**